File: remx-cli/src/remx/commands/relate.py
"""remx relate command — manage memory topology relations."""
from __future__ import annotations

import json
import sys
from pathlib import Path
from typing import Optional

from ..core.topology import (
    DEFAULT_CONTEXT,
    REL_TYPES,
    delete_relation,
    get_related_nodes,
    insert_relation,
    list_nodes,
    query_relations,
    topology_aware_recall,
)


def run_relate(
    db_path: Path,
    action: str,
    *,
    node_id: Optional[str] = None,
    rel_type: Optional[str] = None,
    context: Optional[str] = None,
    description: Optional[str] = None,
    roles: Optional[str] = None,
    current_context: Optional[str] = None,
    max_depth: int = 2,
    max_additional: int = 10,
    limit: int = 50,
) -> int:
    """Manage topology relations between memory entries.

    Actions:
      insert    Insert a new relation
      delete    Delete a relation by ID
      query     Query relations for a node
      nodes     List all nodes (optionally filtered by category)
      graph     BFS traversal to get related nodes
      expand    Expand base semantic results via topology

    Returns:
        0 on success, 1 on error
    """
    if not db_path.exists():
        print(f"remx relate: {db_path}: database not found", file=sys.stderr)
        return 1

    try:
        if action == "nodes":
            nodes = list_nodes(db_path, category=None)
            for n in nodes[:limit]:
                print(f"{n['id']} [{n['category']}] {n['chunk'][:60]}")
            print(f"({len(nodes)} nodes total)", file=sys.stderr)
            return 0

        elif action == "insert":
            if not node_id:
                print("remx relate insert: --node-id required", file=sys.stderr)
                return 1
            if not rel_type:
                print("remx relate insert: --rel-type required", file=sys.stderr)
                return 1
            if rel_type not in REL_TYPES:
                print(f"remx relate insert: invalid rel_type. Options: {', '.join(REL_TYPES)}", file=sys.stderr)
                return 1

            # Parse roles if provided (comma-separated, must match node_id count)
            if roles:
                role_list = [r.strip() for r in roles.split(",")]
            else:
                # Default: first node is cause, rest are effects
                print("remx relate insert: --roles recommended (default: cause for first, effect for rest)", file=sys.stderr)
                role_list = []

            # For now, simplified: single bidirectional relation between two nodes
            # node_id format: id1,id2 (comma-separated)
            node_ids = [n.strip() for n in node_id.split(",")]
            if len(node_ids) < 2:
                print("remx relate insert: need at least 2 node IDs (comma-separated)", file=sys.stderr)
                return 1

            if not role_list:
                role_list = ["cause"] + ["effect"] * (len(node_ids) - 1)

            rel_id = insert_relation(
                db_path=db_path,
                rel_type=rel_type,
                node_ids=node_ids,
                roles=role_list,
                context=context,
                description=description,
            )
            print(f"rel_id={rel_id}")
            return 0

        elif action == "delete":
            if not node_id:
                print("remx relate delete: --node-id required (pass relation ID as integer)", file=sys.stderr)
                return 1
            try:
                rel_id = int(node_id)
            except ValueError:
                print(f"remx relate delete: relation ID must be integer, got: {node_id}", file=sys.stderr)
                return 1
            delete_relation(db_path, rel_id)
            print(f"deleted relation {rel_id}")
            return 0

        elif action == "query":
            if not node_id:
                print("remx relate query: --node-id required", file=sys.stderr)
                return 1
            rels = query_relations(db_path, node_id, current_context)
            print(json.dumps(rels, indent=2, ensure_ascii=False))
            return 0

        elif action == "graph":
            if not node_id:
                print("remx relate graph: --node-id required", file=sys.stderr)
                return 1
            graph = get_related_nodes(db_path, node_id, current_context, max_depth)
            print(json.dumps(graph, indent=2, ensure_ascii=False))
            return 0

        elif action == "expand":
            # For expand, we need base_results passed via stdin
            # base_results: list of entry dicts with at least 'id' field
            try:
                base_raw = sys.stdin.read()
                if not base_raw.strip():
                    base_results = []
                else:
                    base_results = json.loads(base_raw)
            except json.JSONDecodeError as e:
                print(f"remx relate expand: invalid JSON from stdin — {e}", file=sys.stderr)
                return 1

            expanded = topology_aware_recall(
                db_path=db_path,
                base_results=base_results,
                current_context=current_context,
                max_depth=max_depth,
                max_additional=max_additional,
            )
            print(json.dumps(expanded, indent=2, ensure_ascii=False))
            return 0

        else:
            print(f"remx relate: unknown action: {action}", file=sys.stderr)
            print(f"Actions: insert, delete, query, nodes, graph, expand", file=sys.stderr)
            return 1

    except Exception as e:
        print(f"remx relate: {action} error — {e}", file=sys.stderr)
        import traceback
        traceback.print_exc()
        return 1

File: remx-core/src/commands/relate.ts
/**
 * remx relate — topology relation management CLI
 *
 * Actions:
 *   nodes   List all nodes
 *   insert  Insert a new relation
 *   delete  Delete a relation by ID
 *   query   Query relations for a node
 *   graph   BFS traversal to get related nodes
 */
import { Command } from "commander";
import {
  listNodes,
  insertTriple,
  deleteRelation,
  queryRelations,
  getRelatedNodes,
  topologyAwareRecall,
  REL_TYPES_ARRAY,
  type RelType,
  type BaseResult,
} from "../index";

/** Read all available stdin data as a string. */
function readStdin(): Promise<string> {
  return new Promise((resolve) => {
    const chunks: string[] = [];
    process.stdin.on("data", (chunk: Buffer) => chunks.push(chunk.toString()));
    process.stdin.on("end", () => resolve(chunks.join("")));
  });
}

export function makeRelateCommand(): Command {
  const cmd = new Command("relate");
  cmd.description("manage topology relations between memory nodes");

  // ─── nodes ────────────────────────────────────────────────────────────────
  cmd
    .command("nodes")
    .description("list all topology nodes")
    .requiredOption("--db <path>", "path to SQLite database")
    .option("--category <cat>", "filter by category")
    .option("--limit <n>", "max nodes to show", "50")
    .action(async (opts) => {
      const nodes = listNodes(opts.db);
      const filtered = opts.category ? nodes.filter((n) => n.category === opts.category) : nodes;
      const limit = parseInt(opts.limit, 10);
      const shown = filtered.slice(0, limit);
      for (const n of shown) {
        const preview = n.chunk.slice(0, 60).replace(/\n/g, " ");
        console.log(`${n.id} [${n.category}] ${preview}`);
      }
      console.log(`(${filtered.length} nodes total)`);
    });

  // ─── insert ────────────────────────────────────────────────────────────────
  cmd
    .command("insert")
    .description("insert a new topology relation")
    .requiredOption("--db <path>", "path to SQLite database")
    .requiredOption("--nodes <ids>", "comma-separated node IDs (min 2)")
    .requiredOption("--rel-type <type>", `relation type: ${REL_TYPES_ARRAY.join(", ")}`)
    .option(
      "--roles <roles>",
      "comma-separated roles (auto: cause for first, effect for rest)"
    )
    .option("--context <ctx>", "context label (NULL = global)", "global")
    .option("--description <text>", "optional relation description")
    .action(async (opts) => {
      const { db, nodes, relType, roles, context, description } = opts;

      if (!REL_TYPES_ARRAY.includes(relType as RelType)) {
        console.error(`invalid rel-type: ${relType}. Options: ${REL_TYPES_ARRAY.join(", ")}`);
        process.exit(1);
      }

      const nodeIds = nodes.split(",").map((n: string) => n.trim());
      if (nodeIds.length < 2) {
        console.error("need at least 2 node IDs");
        process.exit(1);
      }

      const roleList = roles
        ? roles.split(",").map((r: string) => r.trim())
        : ["cause", ...Array(nodeIds.length - 1).fill("effect")];

      const ctx = context === "global" || context === "null" ? null : context;

      const relId = insertTriple({
        dbPath: db,
        relType: relType as RelType,
        nodeIds,
        roles: roleList as any,
        context: ctx,
        description: description ?? undefined,
      });

      console.log(`rel_id=${relId}`);
    });

  // ─── delete ───────────────────────────────────────────────────────────────
  cmd
    .command("delete")
    .description("delete a relation by ID")
    .requiredOption("--db <path>", "path to SQLite database")
    .requiredOption("--rel-id <id>", "relation ID to delete")
    .action(async (opts) => {
      const id = parseInt(opts.relId, 10);
      if (isNaN(id)) {
        console.error(`invalid relation ID: ${opts.relId}`);
        process.exit(1);
      }
      deleteRelation(opts.db, id);
      console.log(`deleted relation ${id}`);
    });

  // ─── query ────────────────────────────────────────────────────────────────
  cmd
    .command("query")
    .description("query relations for a node")
    .requiredOption("--db <path>", "path to SQLite database")
    .requiredOption("--node-id <id>", "node ID to query")
    .option("--current-context <ctx>", "context for filtering (default: global)", "global")
    .action(async (opts) => {
      const ctx = opts.currentContext === "global" ? null : opts.currentContext;
      const rels = queryRelations(opts.db, opts.nodeId, ctx ?? undefined);
      console.log(JSON.stringify(rels, null, 2));
    });

  // ─── graph ────────────────────────────────────────────────────────────────
  cmd
    .command("graph")
    .description("BFS traversal to get related nodes")
    .requiredOption("--db <path>", "path to SQLite database")
    .requiredOption("--node-id <id>", "starting node ID")
    .option(
      "--current-context <ctx>",
      "context for filtering (default: global)",
      "global"
    )
    .option("--max-depth <n>", "max BFS depth", "2")
    .action(async (opts) => {
      const ctx = opts.currentContext === "global" ? null : opts.currentContext;
      const maxDepth = parseInt(opts.maxDepth, 10);
      const graph = getRelatedNodes(opts.db, opts.nodeId, ctx ?? undefined, maxDepth);
      console.log(JSON.stringify(graph, null, 2));
    });

  // ─── expand ───────────────────────────────────────────────────────────────
  cmd
    .command("expand")
    .description("expand base results via topology graph (reads JSON from stdin)")
    .requiredOption("--db <path>", "path to SQLite database")
    .option(
      "--current-context <ctx>",
      "context for filtering (default: global)",
      "global"
    )
    .option("--max-depth <n>", "max BFS depth", "2")
    .option("--max-additional <n>", "max additional results", "10")
    .option("--stdin", "read base results as JSON array from stdin", false)
    .action(async (opts) => {
      if (!opts.stdin) {
        console.error("remx relate expand: --stdin is required (pipe base results as JSON)");
        process.exit(1);
      }
      const raw = await readStdin();
      let baseResults: BaseResult[];
      try {
        baseResults = (JSON.parse(raw || "[]") as unknown) as BaseResult[];
      } catch {
        console.error("remx relate expand: invalid JSON from stdin");
        process.exit(1);
      }
      const ctx = opts.currentContext === "global" ? null : opts.currentContext;
      const expanded = topologyAwareRecall(baseResults, {
        dbPath: opts.db,
        currentContext: ctx,
        maxDepth: parseInt(opts.maxDepth, 10),
        maxAdditional: parseInt(opts.maxAdditional, 10),
      });
      console.log(JSON.stringify(expanded, null, 2));
    });

  return cmd;
}
